### PR TITLE
NIAD-2394: Use placeholders when mapping DocumentReferences, so that any MigrateDocument failures are represented as AbsentAttachments in XML

### PR DESCRIPTION
**Pull request by adrianclay**
## What

Please include a summary of the changes and the related issue

## Why

Currently, when a 404 or similar happens when fetching a document, the document sent over isn't prefixed with the mandated "AbsentAttachment", neither does the error message get sent, nor the content type updated to text/plain.

This change will mean when an NME fails to serve up a document we can still send over a valid GP2GP extract.

## Type of change

Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have updated the [Changelog](CHANGELOG.md) with details of my change in the UNRELEASED section if this change will affect end users

### CODE REVIEW

**Comment by adrianclay** [comment]
Builds off #736, so merge that before this.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/gpc/GetGpcStructuredTaskExecutor.java [hunk 1]
```diff
@@ -128,14 +129,15 @@ public void execute(GetGpcStructuredTaskDefinition structuredTaskDefinition) {
             messageContext.resetMessageContext();
         }
 
-        var allExternalAttachments = Stream
-            .concat(externalAttachments.stream(), absentAttachments.stream())
-            .collect(Collectors.toList());
+        var allExternalAttachments = Stream.concat(externalAttachments.stream(), absentAttachments.stream())
+            .peek(this::replaceFilenameWithPlaceholder)
+            .peek(this::addPrefixToDocumentId)
+            .toList();
```
Comment (adrianclay) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/gpc/GetGpcStructuredTaskExecutor.java:135]: We also need to be updating the `content_type` field within the JSON ExternalMessage representation, it should be `text/plain` for an AbsentAttachment.

```json
{
    "description": "Filename=\"CC13A057-11A5-435A-913C-731A1670DD88.doc\" ContentType=application/msword Compressed=No LargeAttachment=No OriginalBase64=No Length=236",
    "document_id": "_CC13A057-11A5-435A-913C-731A1670DD88",
    "message_id": "9723E7E1-9FF9-479C-A038-5E0343D485C5"
}
```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/gpc/GetGpcDocumentTaskExecutor.java [hunk 2]
```diff
@@ -94,8 +95,9 @@ private EhrExtractStatus handleValidGpcDocument(String response, GetGpcDocumentT
 
         storageConnectorService.uploadFile(storageDataWrapperWithMhsOutboundRequest, storagePath);
 
+        final var filename = DocumentReferenceUtils.buildPresentAttachmentFileName(taskDefinition.getDocumentId(), binary.getContentType());
         return ehrExtractStatusService.updateEhrExtractStatusAccessDocument(
-            taskDefinition, storagePath, contentAsBase64.length(), null);
+            taskDefinition, storagePath, contentAsBase64.length(), null, filename);
```
Comment (adrianclay) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/gpc/GetGpcDocumentTaskExecutor.java:100]: No test coverage for this change (I think).

File: service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/model/OutboundMessage.java [hunk 3]
```diff
@@ -101,20 +102,24 @@ public static class AttachmentDescription {
         @Override
         public String toString() {
```
Comment (adrianclay) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/mhs/model/OutboundMessage.java:103]: This method hasn't been tested for the `fileName`/`contentType` NULL values.

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractStatusServiceTest.java [hunk 4]
```diff
@@ -63,11 +67,74 @@ public void emptyDatabase() {
...
+
+        Map<String, String> results = ehrExtractStatusService.fetchDocumentObjectNameAndSize(inProgressConversationId);
+
+        Map<String, String> expectedResult = new HashMap<>();
+        expectedResult.put("FILENAME_PLACEHOLDER_ID=4E0C8345-A9AB-48EA-8882-DC9E9F3F5F60", "4E0C8345-A9AB-48EA-8882-DC9E9F3F5F60.rtx");
+        expectedResult.put("LENGTH_PLACEHOLDER_ID=4E0C8345-A9AB-48EA-8882-DC9E9F3F5F60", "244");
+        expectedResult.put("ERROR_MESSAGE_PLACEHOLDER_ID=4E0C8345-A9AB-48EA-8882-DC9E9F3F5F60", "");
+        expectedResult.put("CONTENT_TYPE_PLACEHOLDER_ID=4E0C8345-A9AB-48EA-8882-DC9E9F3F5F60", "application/msword");
+
+        assertThat(results.size()).isEqualTo(4);
```
Comment (ole4ryb) [inline service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractStatusServiceTest.java:129]: I think assertThat(results).hasSize(4); is more concise.

**Review by MartinWheelerMT — APPROVED** [review]

File: service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractStatusServiceTest.java [hunk 5]
```diff
@@ -63,6 +66,58 @@ public void emptyDatabase() {
...
+        var inProgressConversationId = generateRandomUppercaseUUID();
+
+        addInProgressTransfer(
+            inProgressConversationId, List.of(
+                EhrExtractStatus.GpcDocument.builder()
+                    .fileName("AbsentAttachment4E0C8345-A9AB-48EA-8882-DC9E9F3F5F60.rtx")
+                    .documentId("4E0C8345-A9AB-48EA-8882-DC9E9F3F5F60")
+                    .contentLength(DEFAULT_CONTENT_LENGTH)
+                    .gpConnectErrorMessage("404 Not Found")
+                    .contentType("application/msword")
```
Comment (ole4ryb) [inline service/src/intTest/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractStatusServiceTest.java:80]: Minor but I think "application/msword" can be defined as a constant as it is used 3 times in the tests.

**Pull request merged**